### PR TITLE
Expose required fields on form

### PR DESCRIPTION
**Pull request by davi-tapajos-codeminer42**
Displays an asterisk beside label on form when field is required

Issue: https://github.com/thoughtbot/administrate/issues/1489

Captura de Tela 2020-01-09 às 17 20 38

### CODE REVIEW

File: app/helpers/administrate/application_helper.rb [hunk 1]
```diff
@@ -15,6 +15,20 @@ def render_field(field, locals = {})
       render locals: locals, partial: field.to_partial_path
     end
 
+    def requireness(field)
```
Comment (marclerodrigues) [inline app/helpers/administrate/application_helper.rb]: Shouldn't we have specs for these other methods as well?

**Comment by marclerodrigues** [comment]
Overall it looks really good.

File: app/helpers/administrate/application_helper.rb [hunk 2]
```diff
@@ -15,6 +15,20 @@ def render_field(field, locals = {})
...
+      required_field?(field) ? "required" : "optional"
+    end
+
+    def required_field?(field)
+      has_presence_validator?(field.resource.class, field.attribute)
+    end
+
+    def has_presence_validator?(resource_class, field_name)
+      resource_class.validators_on(field_name).
```
Comment (marclerodrigues) [inline app/helpers/administrate/application_helper.rb]: I don't know if using a trailing dot in a multi-line chain is a project standard but I'd go with https://github.com/rubocop-hq/ruby-style-guide#leading-dot-in-multi-line-chains.

Comment (marclerodrigues) [inline app/helpers/administrate/application_helper.rb]: What do you think about doing only one iteration over these validators?
```ruby
def has_presence_validator?(resource_class, field_name)
  resource_class
    .validators_on(field_name)
    .any? { |validator| validator.class == ActiveRecord::Validations::PresenceValidator }
end
```

File: app/helpers/administrate/application_helper.rb (hunk 1 above)
Comment (davitapajos-cm42) [reply app/helpers/administrate/application_helper.rb]: These two methods, in the end, calls "has_presence_validator?" method. If I test "requireness" method I would be testing the Ruby's ternary operator, no? Do you think it is necessary?

File: app/helpers/administrate/application_helper.rb (hunk 2 above)
Comment (davitapajos-cm42) [reply app/helpers/administrate/application_helper.rb]: It is a project standard

Comment (davitapajos-cm42) [reply app/helpers/administrate/application_helper.rb]: Nice!

Comment (davitapajos-cm42) [reply app/helpers/administrate/application_helper.rb]: this line will exceed line size:
```ruby
any? { |validator| validator.class == ActiveRecord::Validations::PresenceValidator }
```

do you think that would be ok?
```ruby
any? { |v| v.class == ActiveRecord::Validations::PresenceValidator }
```

File: app/helpers/administrate/application_helper.rb (hunk 1 above)
Comment (marclerodrigues) [reply app/helpers/administrate/application_helper.rb]: You will be testing the return of your function depending on the params passed to it.

File: app/helpers/administrate/application_helper.rb (hunk 2 above)
Comment (marclerodrigues) [reply app/helpers/administrate/application_helper.rb]: In such a small context it is fine.

File: app/helpers/administrate/application_helper.rb (hunk 1 above)
Comment (pablobm) [reply app/helpers/administrate/application_helper.rb]: I think the provided specs are enough. Testing `requireness` already is a test for the others.

File: spec/helpers/administrate/application_helper_spec.rb [hunk 3]
```diff
@@ -54,6 +54,32 @@
     end
   end
 
+  describe "#requireness" do
+    let(:page) { Administrate::Page::Form.new(Blog::PostDashboard.new, Blog::Post.new) }
+
+    it "returns 'required' if field is required" do
+      title_field = page.attributes.select { |i| i.attribute == :title }.first
```
[the remark this answers is not in the document]
Comment (davitapajos-cm42) [reply spec/helpers/administrate/application_helper_spec.rb]: nice

[the remark this answers is not in the document]
Comment (davi-tapajos-codeminer42) [reply spec/helpers/administrate/application_helper_spec.rb]: Hound bot told me to "prefer detect over find" :)

File: spec/helpers/administrate/application_helper_spec.rb [hunk 4]
```diff
@@ -54,6 +54,34 @@
     end
   end
 
+  describe "#requireness" do
+    let(:page) {
```
[the remark this answers is not in the document]
Comment (davi-tapajos-codeminer42) [reply spec/helpers/administrate/application_helper_spec.rb]: I actually don't know how to solve this. This line will be longer than 80 characters if I put it in one line.

There is a good way to break it and still looks nice?

```ruby
let(:page) { Administrate::Page::Form.new(Blog::PostDashboard.new, Blog::Post.new) }
```

[the remark this answers is not in the document]
Comment (pablobm) [reply spec/helpers/administrate/application_helper_spec.rb]: Sure, you have to use `do...end`:

```ruby
let(:page) do
  Administrate::Page::Form.new(Blog::PostDashboard.new, Blog::Post.new)
end
```

I admit it looks a bit weird, but I think it's a reasonable trade-off for a clear coding style.

**Comment by pablobm** [comment]
Thank you for your contribution @davi-tapajos-codeminer42!

**Pull request merged**